### PR TITLE
modules: Add ACL module

### PR DESCRIPTION
**Pull request by changlan**
- Almost identical to the one in NetBricks
- `established` keyword is not yet implemented

### CODE REVIEW

File: core/modules/acl.cc [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+void ACL::ProcessBatch(bess::PacketBatch *batch) {
+  gate_idx_t out_gates[bess::PacketBatch::kMaxBurst];
+  int cnt = batch->cnt();
+  for (int i = 0; i < cnt; i++) {
+    bess::Packet *pkt = batch->pkts()[i];
+
+    struct ether_hdr *eth = pkt->head_data<struct ether_hdr *>();
+    struct ipv4_hdr *ip = reinterpret_cast<struct ipv4_hdr *>(eth + 1);
+    struct udp_hdr *udp = reinterpret_cast<struct udp_hdr *>(ip + 1);
```
Comment (barath) [inline core/modules/acl.cc]: Do we need a check for options here?  If there are options, then ip+1 won't point at the UDP header.

File: core/utils/ether.h [hunk 2]
```diff
@@ -27,7 +27,7 @@ struct [[gnu::packed]] EthHeader {
 
     void Randomize();
 
-    bool operator == (const Address &o) const {
+    bool operator==(const Address &o) const {
```
Comment (barath) [inline core/utils/ether.h]: Given that it's a small value, would it be faster to treat it as an int and compare the ints?

File: core/utils/ether.h [hunk 3]
```diff
@@ -36,7 +36,7 @@ struct [[gnu::packed]] EthHeader {
       return true;
     }
 
-    bool operator !=(const Address &o) const {
+    bool operator!=(const Address &o) const {
```
Comment (barath) [inline core/utils/ether.h]: Why is != needed?

File: core/utils/ip.h [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+namespace bess {
+namespace utils {
+
+typedef uint32_t IPAddress;
+
+struct CIDRNetwork {
+  CIDRNetwork() = default;
+  explicit CIDRNetwork(const std::string& cidr);
+
+  bool Match(const IPAddress& ip) const { return (addr & mask) == (ip & mask); }
```
Comment (barath) [inline core/utils/ip.h]: Method comment.

File: core/utils/ip.h [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+#include <string>
+
+namespace bess {
+namespace utils {
+
+typedef uint32_t IPAddress;
+
+struct CIDRNetwork {
+  CIDRNetwork() = default;
```
Comment (barath) [inline core/utils/ip.h]: Comments for constructors.

File: core/modules/acl.cc [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+    IPAddress src_ip = ntohl(ip->src_addr);
```
Comment (barath) [inline core/modules/acl.cc]: Would it be faster overall to keep everything network byte ordered?  Then these conversions don't have to happen.

File: core/utils/ip_test.cc [hunk 7]
```diff
@@ -0,0 +1,37 @@
+#include "ip.h"
+
+#include <gtest/gtest.h>
+
+namespace {
+
+using bess::utils::CIDRNetwork;
+
+TEST(IPTest, CIDRInStr) {
```
Comment (barath) [inline core/utils/ip_test.cc]: Test function comments.

File: protobuf/module_msg.proto [hunk 8]
```diff
@@ -131,6 +131,18 @@ message WildcardMatchCommandSetDefaultGateArg {
   uint64 gate = 1;
 }
 
+message ACLArg {
```
Comment (barath) [inline protobuf/module_msg.proto]: Protobuf message/field comments.

**Comment by changlan** [comment]
Fixed. There weren't changes to `ether.h` in this PR except formatting.

File: protobuf/module_msg.proto [hunk 9]
```diff
@@ -133,12 +133,12 @@ message WildcardMatchCommandSetDefaultGateArg {
 
 message ACLArg {
   message Rule {
-    string src_ip = 1;
-    string dst_ip = 2;
-    uint32 src_port = 3;
-    uint32 dst_port = 4;
-    bool established = 5;
-    bool drop = 6;
+    string src_ip = 1;    // Source IP: "192.168.0.1". Wildcard if "".
```
Comment (barath) [inline protobuf/module_msg.proto]: Is the given IP an example?

**Comment by barath** [comment]
Just made this...

out

**Pull request merged**